controller/kubernetes_controller.go

```diff
@@ -0,0 +1,382 @@
+		}
+	}
+	ks.PVStatus = string(pv.Status.Phase)
+
+	var pod *v1.Pod
+	pod, err = kc.getAssociatedPod(ks.Namespace, ks.PVCName)
+	if err != nil {
+		return err
+	}
+	// pod deleted. indicating history data by setting <LastPodRefAt>
+	if ks.PodName != "" && ks.LastPodRefAt == "" && (pod == nil || pod.DeletionTimestamp != nil) {
+		if ks.LastPodRefAt == "" {
+			ks.LastPodRefAt = util.Now()
+		}
+	}
+	if pod != nil {
+		workloadName, workloadType := kc.detectWorkload(pod, ks)
+		ks.PodName = pod.Name
+		ks.PodStatus = string(pod.Status.Phase)
+		ks.WorkloadName = workloadName
+		ks.WorkloadType = workloadType
+		ks.LastPodRefAt = ""
+	}
+
+	// if pv.Status.Phase == v1.VolumePending || pv.Status.Phase == v1.VolumeAvailable, do nothing
+
+	volume, err = kc.ds.UpdateVolume(volume)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (kc *KubernetesController) validatePersistentVolume(pv *v1.PersistentVolume) string {
+	if pv == nil {
+		return ""
+	}
+	// try to get associated Longhorn volume name
+	if pv.Spec.CSI == nil || pv.Spec.CSI.VolumeHandle == "" {
+		return ""
+	}
+	return pv.Spec.CSI.VolumeHandle
+}
+
+func (kc *KubernetesController) enqueuePersistentVolume(pv *v1.PersistentVolume) {
+	key, err := controller.KeyFunc(pv)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("Couldn't get key for object %#v: %v", pv, err))
+		return
+	}
+	kc.queue.AddRateLimited(key)
+	return
+}
+
+func (kc *KubernetesController) enqueuePodChange(pod *v1.Pod) {
+	if _, err := controller.KeyFunc(pod); err != nil {
+		utilruntime.HandleError(fmt.Errorf("Couldn't get key for object %#v: %v", pod, err))
+		return
+	}
+
+	for _, v := range pod.Spec.Volumes {
+		if v.VolumeSource.PersistentVolumeClaim != nil {
+			pvc, err := kc.pvcLister.PersistentVolumeClaims(pod.Namespace).Get(v.VolumeSource.PersistentVolumeClaim.ClaimName)
+			if datastore.ErrorIsNotFound(err) {
+				continue
+			}
+			if err != nil {
+				utilruntime.HandleError(fmt.Errorf("Couldn't get key for object %#v: %v", pvc, err))
+				return
+			}
+			pvName := pvc.Spec.VolumeName
+			if pvName != "" {
+				kc.queue.AddRateLimited(pvName)
+			}
+		}
+	}
+	return
+}
+
+func (kc *KubernetesController) enqueuePVDeletion(pv *v1.PersistentVolume) {
+	if _, err := controller.KeyFunc(pv); err != nil {
+		utilruntime.HandleError(fmt.Errorf("Couldn't get key for object %#v: %v", pv, err))
+		return
+	}
+	if pv.Spec.CSI != nil && pv.Spec.CSI.VolumeHandle != "" {
+		kc.pvToVolumeCache.Store(pv.Name, pv.Spec.CSI.VolumeHandle)
+	}
+	return
+}
+
+func (kc *KubernetesController) cleanupForPVDeletion(pvName string) error {
+	volumeName, ok := kc.pvToVolumeCache.Load(pvName)
+	if !ok {
+		return nil
+	}
+	volume, err := kc.ds.GetVolume(volumeName.(string))
+	if err != nil {
+		if datastore.ErrorIsNotFound(err) {
+			logrus.Infof("Longhorn volume %v has been deleted", volumeName)
+			return nil
+		}
+		return err
+	}
+	pv, err := kc.kubeClient.CoreV1().PersistentVolumes().Get(pvName, metav1.GetOptions{})
```

Comment:
Let's add the PVLister as well. Seems we're calling the API for PV at least twice per loop.